step/src/main/java/com/google/sps/MatchManager.java

```diff
@@ -17,50 +17,113 @@
+    return new HashSet <User>(matchQueue);
   }
 
-  // FIFO matchmaker -- takes in the user who just requested a match and matches them
-  // with anyone else who is in the queue waiting to be matched or adds them to the
+  // Setter method to add users directly to the matchqueue (for testing purposes).
+  public static void addUserToMatchQueue(User u) {
+      matchQueue.add(u);
+  }
+
+  // Matchmaker -- takes in the user who just requested a match and matches them
+  // with the one who is in the queue having the highest number of mutual interests with the person
+  // or adds them to the
   // queue to wait to be matched if no on else is in there
   public static void generateMatch(User user) {
     user.updateMatchPendingStatus(true);
 
     if (matchQueue.isEmpty()) {
-      matchQueue.add(user);  
-    } else {
-      User matchResult = matchQueue.peek();
+        matchQueue.add(user); 
+        return;
+    }
+    matchQueue.add(user); 
+
+    if (!matchQueue.isEmpty()) {
+      matchQueue.remove(user);
+      User matchResult = findCompatibleMatch(user);
+      if (matchResult == null) {
+          matchQueue.add(user); 
+          user.updateMatchPendingStatus(true);
+          return;
+      }
 
       if (!matchResult.isMatchedWith(user) && !matchResult.equals(user)) {
         // actually removing the match result from the queue
-        matchResult = matchQueue.poll(); 
         createMatch(matchResult, user);
+        matchQueue.remove(matchResult);  
+        
       // adding the user to the queue in case there wasn't a successful match for them
       // this is so they can be matched later      
-      } else if (!matchQueue.contains(user)) {
-        matchQueue.add(user);  
+      } else if (!matchQueue.contains(user)) {  
+          matchQueue.add(user); 
       }  
     }    
   }
 
+  // Match the user with the most compatible user in the database (the one who has the most number
+  // of similar interests). 
+  // Currently, the minimum percentage of mutual interests between 2 users required to get matched are 60%
+  public static User findCompatibleMatch(User firstUser) {
+    HashMap<User, Integer> mutualInterests = new HashMap<>();
+    for (User secondUser: matchQueue) {
+        mutualInterests.put(secondUser, 0);
+    } 
+    for (User secondUser: matchQueue) {
+        for (int i = 0; i < secondUser.getPreferences().size(); i++) {
+            // Compare the second user's each preference answer with the first user's
+            if (firstUser.getPreferences().get(i).equals(secondUser.getPreferences().get(i))) {
+                mutualInterests.replace(secondUser, mutualInterests.get(secondUser)+1);
+            }
+        }
+    }
+
+    //Traverse through the hashmap to find the user with the maximum number of mutual interests with the first user
+    Iterator iterator = mutualInterests.entrySet().iterator();
+    int max = 0; 
+    String compatibleUserId = "";
+
+    while (iterator.hasNext()) {
```

Comment:
One way to optimize this could be to add users to a priority queue where they're ordered based on their number of mutual interests, that way you can just return whoever is at the top of the queue.

Reply:
If you don't want to deal with a priority queue you could also just keep track of who the current highest scorer is as you loop through using variables like currHighestScore and currHighestUser or something like that.

Reply:
I am not sure if creating a priority queue is necessary but yes, I can definitely change the variable names to make them more intuitive